### PR TITLE
Refactor GitOps approach prompt flow in guided init

### LINKED ISSUE #2268: [BUG] - Nebari init fails when using a GitLab repo for the GitOps approach
### Describe the bug

When running `nebari init` and specifying a GitLab repo, the command fails with the following error:
```python
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /usr/local/bin/nebari:8 in <module>                                                              │
│                                                                                                  │
│   5 from nebari.__main__ import main                                                             │
│   6 if __name__ == '__main__':                                                                   │
│   7 │   sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])                         │
│ ❱ 8 │   sys.exit(main())                                                                         │
│   9                                                                                              │
│                                                                                                  │
│ /usr/local/lib/python3.11/site-packages/nebari/__main__.py:8 in main                             │
│                                                                                                  │
│    5                                                                                             │
│    6 def main():                                                                                 │
│    7 │   cli = create_cli()                                                                      │
│ ❱  8 │   cli()                                                                                   │
│    9                                                                                             │
│   10                                                                                             │
│   11 # get the click object from the typer app so that we can autodoc the cli                    │
│                                                                                                  │
│ /usr/local/lib/python3.11/site-packages/typer/main.py:328 in __call__                            │
│                                                                                                  │
│ /usr/local/lib/python3.11/site-packages/typer/main.py:311 in __call__                            │
│                                                                                                  │
│ /usr/local/lib/python3.11/site-packages/click/core.py:1130 in __call__                           │
│                                                                                                  │
│ /usr/local/lib/python3.11/site-packages/typer/core.py:778 in main                                │
│                                                                                                  │
│ /usr/local/lib/python3.11/site-packages/typer/core.py:216 in _main                               │
│                                                                                                  │
│ /usr/local/lib/python3.11/site-packages/click/core.py:1655 in invoke                             │
│                                                                                                  │
│ /usr/local/lib/python3.11/site-packages/click/core.py:920 in make_context                        │
│                                                                                                  │
│ /usr/local/lib/python3.11/site-packages/click/core.py:1378 in parse_args                         │
│                                                                                                  │
│ /usr/local/lib/python3.11/site-packages/click/core.py:2360 in handle_parse_result                │
│                                                                                                  │
│ /usr/local/lib/python3.11/site-packages/click/core.py:2322 in process_value                      │
│                                                                                                  │
│ /usr/local/lib/python3.11/site-packages/typer/main.py:994 in wrapper                             │
│                                                                                                  │
│ /usr/local/lib/python3.11/site-packages/_nebari/subcommands/init.py:797 in guided_init_wizard    │
│                                                                                                  │
│   794 │   │   │   │   qmark=qmark,                                                               │
│   795 │   │   │   ).unsafe_ask()                                                                 │
│   796 │   │   │                                                                                  │
│ ❱ 797 │   │   │   inputs.repository = repo_url.format(                                           │
│   798 │   │   │   │   git_provider=git_provider, org_name=org_name, repo_name=repo_name          │
│   799 │   │   │   )                                                                              │
│   800                                                                                            │
│                                                                                                  │
│ in pydantic.main.BaseModel.__setattr__:384                                                       │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
ValidationError: 1 validation error for InitInputs
repository
  string does not match regex "^(https://)?github.com/([^/]+)/([^/]+)/?$" (type=value_error.str.regex; pattern=^(https://)?github.com/([^/]+)/([^/]+)/?$)
```

### Expected behavior

I expect `nebari init` to work properly and create the configuration file according to the options I selected.

### OS and architecture in which you are running Nebari

Linux ip-172-31-3-104 6.2.0-1017-aws #17~22.04.1-Ubuntu SMP Fri Nov 17 21:07:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### How to Reproduce the problem?

Go over the `nebari init --guided-init` and in the GitOps section try the following:

```
🪴  This next section is optional but recommended. If you wish to adopt a GitOps approach to managing this platform, we will walk you through a set of questions to get that 
setup. With this setup, Nebari will use GitHub Actions workflows (or GitLab equivalent) to automatically handle the future deployments of your infrastructure.


   Would you like to adopt a GitOps approach to managing Nebari? Yes
   Which git provider would you like to use? gitlab.com
   Which user or organization will this repository live under? (https://gitlab.com/<org-name>/) some-org
   And what will the name of this repository be? (https://gitlab.com/some-org/<repo-name>) some-repo
```

### Command output

_No response_

### Versions and dependencies used.

`nebari==2024.1.1`

### Compute environment

None

### Integrations

_No response_

### Anything else?

@kalpanachinnappan originally found about this issue

### PR DESCRIPTION
**Pull request by marcelovilla**
## Reference Issues or PRs
Closes #2268 

## What does this implement/fix?

_Put a `x` in the boxes that apply_

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features not to work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

## Testing

- [x] Did you test the pull request locally?
- [ ] Did you add new tests?

## How to test this PR
While we have unit tests covering the functionality changes in this PR, I present three different paths a reviewer might want to test manually.

#### Choose GitLab during the GitOps approach prompts
1. Run `nebari init --guided-init`
2. Follow through with any options and select `y` when prompted with `Would you like to adopt a GitOps approach to managing Nebari? (y/N)`
3. Choose `gitlab.com`

You should be prompted with further configuration options. Follow through selecting `N` for the following questions.

4. Run `nebari render -c nebari-config.yaml`

You should see a `.gitlab-ci.yml` file.

#### Choose GitHub during the GitOps approach prompts
1. Run `nebari init --guided-init`
2. Follow through with any options and select `y` when prompted with `Would you like to adopt a GitOps approach to managing Nebari? (y/N)`
3. Choose `github.com`
4. Choose `N` when prompted with `Would you like nebari to create a remote repository on github.com?`

You should be prompted with further configuration options. Follow through selecting `N` for the following questions.

5. Run `nebari render -c nebari-config.yaml`

You should see a `.github/workflows` folder.

#### Choose GitHub with auto-provisioning during the GitOps approach prompts
1. Run `nebari init --guided-init`
2. Follow through with any options and select `y` when prompted with `Would you like to adopt a GitOps approach to managing Nebari? (y/N)`
3. Choose `github.com`
4. Choose `y` when prompted with `Would you like nebari to create a remote repository on github.com?`
5. Pass an organization you have access to
6. Pass a repo name to be created
7. Paste your github username and token if necessary

You should be prompted with further configuration options. Follow through selecting `N` for the following questions.

5. Run `nebari render -c nebari-config.yaml`

You should see a `.github/workflows` folder and there should be newly created GitHub repo under the organization you specified.

### CODE REVIEW

**Comment by marcelovilla** [comment]
Unit tests are failing. I'm looking into it.

File: src/nebari/schema.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 email_regex = "^[^ @]+@[^ @]+\\.[^ @]+$"
 email_pydantic = pydantic.constr(regex=email_regex)
 
-github_url_regex = "^(https://)?github.com/([^/]+)/([^/]+)/?$"
+github_url_regex = "^(https://)?(github.com|gitlab.com)/([^/]+)/([^/]+)/?$"
```
Comment (aktech) [inline src/nebari/schema.py:20]: How do you feel about writing something like:

```python
git_repo_url_regex = "^(https://)?(git)(hub|lab)(.com)/([^/]+)/([^/]+)/?$"
```

- to remove repetition of common string `git` and `.com`
- also it's no more just github :)

**Review by aktech — CHANGES_REQUESTED** [review]
I have little suggestion for improving the regex. I would also suggest writing some tests, like for example:

```python
import re
import pytest

@pytest.mark.parametrize("url", [
    "https://github.com/user/repository",
    "https://gitlab.com/user/repository/",
])
def test_valid_git_repo_urls(url):
    assert re.match(git_repo_url_regex, url) is not None, f"Failed to match {url}"

@pytest.mark.parametrize("url", [
    "https://github.com",
    "https://github.com/user",
    "https://gitlab.com/user",
    "http://github.com/user/",
    "gitlab.com/user/",
    "github.com",
])
def test_invalid_git_repo_urls(url):
    assert re.match(git_repo_url_regex, url) is None, f"Unexpected match for {url}"

```

File: src/nebari/schema.py (hunk 1 above)
Comment (marcelovilla) [reply src/nebari/schema.py]: Thanks, Amit! That is indeed much nicer. I just made a slight modification at the end to avoid allowing urls with a trailing slash that do not have a repo name. For example:
```
https://gitlab.com/user/
```

This was not being caught before.

**Comment by marcelovilla** [comment]
@aktech I modified the existing tests and added new ones based on your suggestion. I should have added new tests in the first place so thanks for bringing that up!

File: tests/tests_unit/test_cli_init_repository.py [hunk 2]
```diff
@@ -174,21 +174,45 @@ def test_cli_init_error_repository_missing_env(monkeypatch: pytest.MonkeyPatch):
...
+
+    with tempfile.TemporaryDirectory() as tmp:
+        tmp_file = Path(tmp).resolve() / "nebari-config.yaml"
+        assert tmp_file.exists() is False
 
+        result = runner.invoke(app, args + ["--output", tmp_file.resolve()])
+
+        assert 0 == result.exit_code
+        assert not result.exception
+        assert tmp_file.exists() is True
```
Comment (aktech) [inline tests/tests_unit/test_cli_init_repository.py:197]: This code is very similar to the one below, how do you feel about creating a function and reusing that?

**Comment by dcmcand** [comment]
@marcelovilla can you update the status here?

**Comment by dcmcand** [comment]
@marcelovilla what is the status of this PR? Are you still working on it?

**Comment by marcelovilla** [comment]
This PR was stale for a very long time but I've just added some changes to get it across the finish line. You'll see that I decided to revert the regex changes I wrote before (and thus, also decided to change the title of the PR). The reason behind this is that I did not see any point in modifying the regex to allow both GitHub and GitLab repos when we don't really support repo auto-provisioning for GitLab.

I decided to keep only the regex for GitHub repo validation and instead change the flow in which we present the user with the GitOps approach prompts during the guided init. Previously, we were prompting the user for the following inputs:

1. Provider (i.e., GitHub or GitLab)
2. Organization and repository
3. Whether they want Nebari to auto-provision the repository

When users selected GitLab for (1) and then gave an org and repo name for (2), they would get a validation error because of the repo URL (see #2268). While having a regex that validated both provider repos would have solved that issue, it wouldn't really have done anything with that information afterwards, because, as I mentioned before, we have not implemented the auto-provisioning logic for GitLab. I think implementing that is outside of the scope of this PR and will require a considerably larger effort to achieve. Furthermore, in the case where users choose GitHub as the provider, we only use the organization and repository name if they want Nebari to auto-provision the repository. Thus, with this PR my proposal is to modify the GitOps approach flow in the guided init to prompt the user for the following inputs:

1. Provider (i.e., GitHub or GitLab)
2. (Only if the provider is GitHub) whether they want Nebari to auto-provision the repository
3. (Only if they want to auto-provision the repo) organization and repository

Additionally, I modified our render tests to make sure we're passing GitLab as a provider in one of the configurations. And unrelated to this, I fixed the original regex to avoid matching an edge case, which I added to the existing tests.

The PR description has a section on how to test this PR.

**Review by Adam-D-Lewis — APPROVED** [review]
LGTM

**Comment by marcelovilla** [comment]
Thanks for the review @Adam-D-Lewis

@aktech I just realized I cannot merge this as you requested changes a while ago. Can you have a look and see what you think?

**Comment by marcelovilla** [comment]
@aktech just wanted to give this PR a nudge when you have some time. Thanks!

**Review by aktech — APPROVED** [review]
Looks good! 👍
